File: backend/portfolio/risk.py
"""
Portfolio risk analytics — VaR, CVaR, drawdown, Sharpe, Sortino, beta,
correlation matrix, and sector concentration.
"""

import logging
from datetime import date, timedelta
from typing import Optional

import pandas as pd
import numpy as np
from sqlalchemy.orm import Session
from sqlalchemy import func

from config import get_settings
from db.database import SessionLocal
from db.models import (
    PortfolioPosition, PortfolioSnapshot, StockPrice,
    MacroIndicator, RegimeState,
)

logger = logging.getLogger(__name__)
settings = get_settings()

TRADING_DAYS_PER_YEAR = 252


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _get_portfolio_returns(session: Session, days: int = 252) -> pd.Series:
    """
    Build a daily portfolio return series from historical stock prices,
    weighted by each position's market value.

    Returns:
        pd.Series indexed by date with daily portfolio returns.
    """
    positions = (
        session.query(PortfolioPosition)
        .all()
    )
    if not positions:
        return pd.Series(dtype=float)

    start_date = date.today() - timedelta(days=int(days * 1.5))

    # Collect per-ticker weights (market-value weight based on entry)
    ticker_qty: dict[str, float] = {}
    ticker_entry_price: dict[str, float] = {}
    for p in positions:
        t = p.ticker
        ticker_qty[t] = ticker_qty.get(t, 0.0) + p.quantity
        # Use entry_price as a fallback weight base
        ticker_entry_price.setdefault(t, p.entry_price)

    tickers = list(ticker_qty.keys())

    # Fetch prices from DB
    prices_q = (
        session.query(StockPrice.ticker, StockPrice.date, StockPrice.adj_close)
        .filter(
            StockPrice.ticker.in_(tickers),
            StockPrice.date >= start_date,
        )
        .order_by(StockPrice.date)
        .all()
    )

    if not prices_q:
        return pd.Series(dtype=float)

    # Pivot into a DataFrame: rows=date, columns=ticker
    df = pd.DataFrame(prices_q, columns=["ticker", "date", "adj_close"])
    price_matrix = df.pivot(index="date", columns="ticker", values="adj_close")
    price_matrix.sort_index(inplace=True)
    price_matrix.dropna(how="all", inplace=True)
    price_matrix.ffill(inplace=True)

    # Daily returns per ticker
    returns_matrix = price_matrix.pct_change().dropna(how="all")

    # Compute market-value weights
    total_value = sum(
        ticker_qty[t] * (price_matrix[t].iloc[-1] if t in price_matrix.columns else ticker_entry_price[t])
        for t in tickers
        if t in price_matrix.columns
    )

    if total_value == 0:
        return pd.Series(dtype=float)

    weights: dict[str, float] = {}
    for t in tickers:
        if t not in price_matrix.columns:
            continue
        mv = ticker_qty[t] * price_matrix[t].iloc[-1]
        weights[t] = mv / total_value

    # Weighted portfolio return
    portfolio_returns = pd.Series(0.0, index=returns_matrix.index)
    for t, w in weights.items():
        if t in returns_matrix.columns:
            portfolio_returns += returns_matrix[t].fillna(0.0) * w

    # Trim to requested number of trading days
    portfolio_returns = portfolio_returns.iloc[-days:]
    return portfolio_returns


def _compute_var(returns: pd.Series, confidence: float = 0.95) -> float:
    """
    Historical simulation VaR.

    Returns the loss threshold at the given confidence level (as a positive
    number representing percent loss).
    """
    if returns.empty:
        return 0.0
    cutoff = returns.quantile(1 - confidence)
    return float(-cutoff)


# ---------------------------------------------------------------------------
# Main risk computation
# ---------------------------------------------------------------------------

def compute_risk_metrics(session: Session) -> dict:
    """
    Compute a comprehensive set of portfolio risk metrics.

    Returns:
        dict with keys: var_95, var_99, cvar_95, cvar_99, max_drawdown,
        sharpe_ratio, sortino_ratio, beta, correlation_matrix,
        sector_concentration_hhi.
    """
    returns = _get_portfolio_returns(session, days=TRADING_DAYS_PER_YEAR)

    result: dict = {}

    if returns.empty or len(returns) < 5:
        logger.warning("Insufficient return data for risk metrics")
        return result

    # ---- VaR ----
    var_95 = _compute_var(returns, 0.95)
    var_99 = _compute_var(returns, 0.99)
    result["var_95"] = round(var_95, 6)
    result["var_99"] = round(var_99, 6)

    # ---- CVaR (Expected Shortfall) ----
    cutoff_95 = returns.quantile(0.05)
    tail_95 = returns[returns <= cutoff_95]
    result["cvar_95"] = round(float(-tail_95.mean()), 6) if not tail_95.empty else var_95

    cutoff_99 = returns.quantile(0.01)
    tail_99 = returns[returns <= cutoff_99]
    result["cvar_99"] = round(float(-tail_99.mean()), 6) if not tail_99.empty else var_99

    # ---- Max drawdown ----
    cumulative = (1 + returns).cumprod()
    running_max = cumulative.cummax()
    drawdowns = (cumulative - running_max) / running_max
    result["max_drawdown"] = round(float(drawdowns.min()), 6)

    # ---- Risk-free rate from macro_indicators (fed_funds_rate) ----
    risk_free_annual = 0.0
    latest_ffr = (
        session.query(MacroIndicator)
        .filter(MacroIndicator.indicator_name == "fed_funds_rate")
        .order_by(MacroIndicator.date.desc())
        .first()
    )
    if latest_ffr and latest_ffr.value is not None:
        # The stored value is typically in percent (e.g. 5.33 for 5.33%)
        risk_free_annual = latest_ffr.value / 100.0

    risk_free_daily = risk_free_annual / TRADING_DAYS_PER_YEAR

    # ---- Sharpe ratio (annualized) ----
    excess = returns - risk_free_daily
    if returns.std() != 0:
        sharpe = (excess.mean() / returns.std()) * np.sqrt(TRADING_DAYS_PER_YEAR)
        result["sharpe_ratio"] = round(float(sharpe), 4)
    else:
        result["sharpe_ratio"] = 0.0

    # ---- Sortino ratio (annualized) ----
    downside = returns[returns < risk_free_daily] - risk_free_daily
    downside_std = np.sqrt((downside ** 2).mean()) if len(downside) > 0 else 0.0
    if downside_std != 0:
        sortino = (excess.mean() / downside_std) * np.sqrt(TRADING_DAYS_PER_YEAR)
        result["sortino_ratio"] = round(float(sortino), 4)
    else:
        result["sortino_ratio"] = 0.0

    # ---- Beta vs S&P 500 ----
    result["beta"] = _compute_beta(session, returns)

    # ---- Correlation matrix between holdings ----
    result["correlation_matrix"] = _correlation_matrix(session)

    # ---- Sector concentration (Herfindahl-Hirschman Index) ----
    result["sector_concentration_hhi"] = _sector_hhi(session)

    return result


# ---------------------------------------------------------------------------
# Sub-computations
# ---------------------------------------------------------------------------

def _compute_beta(session: Session, portfolio_returns: pd.Series) -> Optional[float]:
    """Compute portfolio beta relative to S&P 500 (^GSPC / SPY)."""
    if portfolio_returns.empty:
        return None

    start_date = date.today() - timedelta(days=int(TRADING_DAYS_PER_YEAR * 1.5))

    # Try SPY as the benchmark
    bench_prices = (
        session.query(StockPrice.date, StockPrice.adj_close)
        .filter(
            StockPrice.ticker == "SPY",
            StockPrice.date >= start_date,
        )
        .order_by(StockPrice.date)
        .all()
    )

    if not bench_prices or len(bench_prices) < 10:
        return None

    bench_df = pd.DataFrame(bench_prices, columns=["date", "adj_close"])
    bench_df.set_index("date", inplace=True)
    bench_returns = bench_df["adj_close"].pct_change().dropna()

    # Align
    common = portfolio_returns.index.intersection(bench_returns.index)
    if len(common) < 10:
        return None

    pr = portfolio_returns.loc[common]
    br = bench_returns.loc[common]

    cov = np.cov(pr, br)
    if cov[1, 1] == 0:
        return None

    beta = cov[0, 1] / cov[1, 1]
    return round(float(beta), 4)


def _correlation_matrix(session: Session) -> dict:
    """
    Return a correlation matrix between all current holdings as a nested dict.
    """
    positions = session.query(PortfolioPosition).all()
    tickers = list({p.ticker for p in positions})

    if len(tickers) < 2:
        return {}

    start_date = date.today() - timedelta(days=int(TRADING_DAYS_PER_YEAR * 1.5))

    prices_q = (
        session.query(StockPrice.ticker, StockPrice.date, StockPrice.adj_close)
        .filter(
            StockPrice.ticker.in_(tickers),
            StockPrice.date >= start_date,
        )
        .order_by(StockPrice.date)
        .all()
    )

    if not prices_q:
        return {}

    df = pd.DataFrame(prices_q, columns=["ticker", "date", "adj_close"])
    price_matrix = df.pivot(index="date", columns="ticker", values="adj_close")
    price_matrix.sort_index(inplace=True)
    price_matrix.ffill(inplace=True)

    returns_matrix = price_matrix.pct_change().dropna(how="all")
    corr = returns_matrix.corr()

    # Convert to a JSON-friendly nested dict with rounded values
    return {
        t1: {t2: round(float(corr.loc[t1, t2]), 4) for t2 in corr.columns}
        for t1 in corr.index
    }


def _sector_hhi(session: Session) -> float:
    """
    Herfindahl-Hirschman Index of sector concentration.

    HHI ranges from 0 (perfectly diversified) to 10000 (single sector).
    """
    positions = session.query(PortfolioPosition).all()
    if not positions:
        return 0.0

    sector_values: dict[str, float] = {}
    total = 0.0
    for p in positions:
        price = p.current_price if p.current_price else p.entry_price
        mv = price * p.quantity
        sector = p.sector or "Unknown"
        sector_values[sector] = sector_values.get(sector, 0.0) + mv
        total += mv

    if total == 0:
        return 0.0

    hhi = sum((v / total * 100) ** 2 for v in sector_values.values())
    return round(hhi, 2)

File: backend/src/portfolio/risk.ts
/**
 * Portfolio risk analytics — VaR, CVaR, drawdown, Sharpe, Sortino, beta,
 * correlation matrix, concentration, clustering, and stress testing.
 */

import { PrismaClient } from '@prisma/client';
import { getPrisma } from '../db/client';
import {
  mean, std, percentile, quantile, dailyReturns, correlation,
  correlationMatrix as buildCorrelationMatrix, round,
} from '../utils/math';
import { daysAgo } from '../utils/format';
import { fetchPrices } from '../ingestion/yahoo';

const TRADING_DAYS_PER_YEAR = 252;

async function getPortfolioReturns(db: PrismaClient, days: number = 252): Promise<{ dates: Date[]; returns: number[] }> {
  const positions = await db.portfolioPosition.findMany();
  if (positions.length === 0) return { dates: [], returns: [] };

  const startDate = daysAgo(Math.floor(days * 1.5));
  const tickerQty: Record<string, number> = {};
  const tickerEntryPrice: Record<string, number> = {};

  for (const p of positions) {
    const t = p.ticker;
    tickerQty[t] = (tickerQty[t] || 0) + Number(p.quantity);
    if (!tickerEntryPrice[t]) tickerEntryPrice[t] = Number(p.entryPrice);
  }

  const tickers = Object.keys(tickerQty);

  const pricesRaw = await db.stockPrice.findMany({
    where: { ticker: { in: tickers }, date: { gte: startDate } },
    orderBy: { date: 'asc' },
    select: { ticker: true, date: true, adjClose: true },
  });

  if (pricesRaw.length === 0) return { dates: [], returns: [] };

  // Build price matrix: date -> ticker -> price
  const dateMap = new Map<string, Record<string, number>>();
  const allDates: string[] = [];

  for (const p of pricesRaw) {
    const dateKey = p.date.toISOString().split('T')[0];
    if (!dateMap.has(dateKey)) {
      dateMap.set(dateKey, {});
      allDates.push(dateKey);
    }
    if (p.adjClose !== null) {
      dateMap.get(dateKey)![p.ticker] = p.adjClose;
    }
  }

  allDates.sort();

  // Forward fill and compute returns
  const prev: Record<string, number> = {};
  const dateReturns: { date: string; ret: number }[] = [];

  for (let i = 0; i < allDates.length; i++) {
    const dateKey = allDates[i];
    const prices = dateMap.get(dateKey)!;

    // Forward fill
    for (const t of tickers) {
      if (prices[t] !== undefined) prev[t] = prices[t];
      else if (prev[t] !== undefined) prices[t] = prev[t];
    }

    if (i === 0) continue;

    const prevDateKey = allDates[i - 1];
    const prevPrices = dateMap.get(prevDateKey)!;

    // Compute total values
    let totalValueNow = 0;
    let totalValuePrev = 0;
    for (const t of tickers) {
      if (prices[t] !== undefined && prevPrices[t] !== undefined) {
        totalValueNow += tickerQty[t] * prices[t];
        totalValuePrev += tickerQty[t] * prevPrices[t];
      }
    }

    if (totalValuePrev > 0) {
      dateReturns.push({ date: dateKey, ret: (totalValueNow - totalValuePrev) / totalValuePrev });
    }
  }

  const trimmed = dateReturns.slice(-days);
  return {
    dates: trimmed.map(d => new Date(d.date)),
    returns: trimmed.map(d => d.ret),
  };
}

function computeVar(returns: number[], confidence: number = 0.95): number {
  if (returns.length === 0) return 0;
  const cutoff = quantile(returns, 1 - confidence);
  return -cutoff;
}

export async function computeRiskMetrics(prisma?: PrismaClient): Promise<Record<string, any>> {
  const db = prisma || getPrisma();
  const { returns } = await getPortfolioReturns(db, TRADING_DAYS_PER_YEAR);
  const result: Record<string, any> = {};

  if (returns.length < 5) return result;

  // VaR
  result.var_95 = round(computeVar(returns, 0.95), 6);
  result.var_99 = round(computeVar(returns, 0.99), 6);

  // CVaR
  const cutoff95 = quantile(returns, 0.05);
  const tail95 = returns.filter(r => r <= cutoff95);
  result.cvar_95 = tail95.length > 0 ? round(-mean(tail95), 6) : result.var_95;

  const cutoff99 = quantile(returns, 0.01);
  const tail99 = returns.filter(r => r <= cutoff99);
  result.cvar_99 = tail99.length > 0 ? round(-mean(tail99), 6) : result.var_99;

  // Max drawdown
  let peak = 1;
  let maxDd = 0;
  let cum = 1;
  for (const r of returns) {
    cum *= (1 + r);
    if (cum > peak) peak = cum;
    const dd = (cum - peak) / peak;
    if (dd < maxDd) maxDd = dd;
  }
  result.max_drawdown = round(maxDd, 6);

  // Risk-free rate
  let riskFreeAnnual = 0;
  const latestFfr = await db.macroIndicator.findFirst({
    where: { indicatorName: 'fed_funds_rate' },
    orderBy: { date: 'desc' },
  });
  if (latestFfr?.value !== undefined) {
    riskFreeAnnual = latestFfr.value / 100;
  }
  const riskFreeDaily = riskFreeAnnual / TRADING_DAYS_PER_YEAR;

  // Sharpe ratio
  const excess = returns.map(r => r - riskFreeDaily);
  const retStd = std(returns, 1);
  if (retStd !== 0) {
    result.sharpe_ratio = round((mean(excess) / retStd) * Math.sqrt(TRADING_DAYS_PER_YEAR), 4);
  } else {
    result.sharpe_ratio = 0;
  }

  // Sortino ratio
  const downside = returns.filter(r => r < riskFreeDaily).map(r => r - riskFreeDaily);
  const downsideStd = downside.length > 0 ? Math.sqrt(mean(downside.map(d => d * d))) : 0;
  if (downsideStd !== 0) {
    result.sortino_ratio = round((mean(excess) / downsideStd) * Math.sqrt(TRADING_DAYS_PER_YEAR), 4);
  } else {
    result.sortino_ratio = 0;
  }

  // Beta
  result.beta = await computeBeta(db, returns);

  // Correlation matrix
  result.correlation_matrix = await buildCorrMatrix(db);

  // Sector HHI
  result.sector_concentration_hhi = await sectorHhi(db);

  return result;
}

async function computeBeta(db: PrismaClient, portfolioReturns: number[]): Promise<number | null> {
  if (portfolioReturns.length === 0) return null;

  const startDate = daysAgo(Math.floor(TRADING_DAYS_PER_YEAR * 1.5));
  let benchPrices = await db.stockPrice.findMany({
    where: { ticker: 'SPY', date: { gte: startDate } },
    orderBy: { date: 'asc' },
    select: { date: true, adjClose: true },
  });

  // Auto-backfill SPY if insufficient data for beta calculation
  if (benchPrices.length < 10) {
    try {
      await fetchPrices('SPY', '2y', db);
      benchPrices = await db.stockPrice.findMany({
        where: { ticker: 'SPY', date: { gte: startDate } },
        orderBy: { date: 'asc' },
        select: { date: true, adjClose: true },
      });
    } catch {
      // Fetch failed — SPY data unavailable
    }
  }

  if (benchPrices.length < 10) return null;

  const benchCloses = benchPrices.filter(p => p.adjClose !== null).map(p => p.adjClose!);
  const benchReturns = dailyReturns(benchCloses);

  const minLen = Math.min(portfolioReturns.length, benchReturns.length);
  if (minLen < 10) return null;

  const pr = portfolioReturns.slice(-minLen);
  const br = benchReturns.slice(-minLen);

  const prMean = mean(pr);
  const brMean = mean(br);
  let cov = 0, brVar = 0;
  for (let i = 0; i < minLen; i++) {
    cov += (pr[i] - prMean) * (br[i] - brMean);
    brVar += (br[i] - brMean) ** 2;
  }

  if (brVar === 0) return null;
  return round(cov / brVar, 4);
}

async function buildCorrMatrix(db: PrismaClient): Promise<Record<string, Record<string, number>>> {
  const positions = await db.portfolioPosition.findMany();
  const tickers = [...new Set(positions.map(p => p.ticker))];

  if (tickers.length < 2) return {};

  const startDate = daysAgo(Math.floor(TRADING_DAYS_PER_YEAR * 1.5));
  const pricesRaw = await db.stockPrice.findMany({
    where: { ticker: { in: tickers }, date: { gte: startDate } },
    orderBy: { date: 'asc' },
    select: { ticker: true, date: true, adjClose: true },
  });

  // Build price series per ticker
  const pricesByTicker: Record<string, number[]> = {};
  for (const t of tickers) pricesByTicker[t] = [];

  const dateMap = new Map<string, Record<string, number>>();
  for (const p of pricesRaw) {
    if (p.adjClose === null) continue;
    const dk = p.date.toISOString().split('T')[0];
    if (!dateMap.has(dk)) dateMap.set(dk, {});
    dateMap.get(dk)![p.ticker] = p.adjClose;
  }

  const dates = [...dateMap.keys()].sort();
  for (const dk of dates) {
    const prices = dateMap.get(dk)!;
    for (const t of tickers) {
      if (prices[t] !== undefined) pricesByTicker[t].push(prices[t]);
    }
  }

  // Compute returns
  const returnsByTicker: Record<string, number[]> = {};
  for (const t of tickers) {
    returnsByTicker[t] = dailyReturns(pricesByTicker[t]);
  }

  const { matrix } = buildCorrelationMatrix(returnsByTicker);

  const result: Record<string, Record<string, number>> = {};
  for (let i = 0; i < tickers.length; i++) {
    result[tickers[i]] = {};
    for (let j = 0; j < tickers.length; j++) {
      result[tickers[i]][tickers[j]] = round(matrix[i][j], 4);
    }
  }

  return result;
}

async function sectorHhi(db: PrismaClient): Promise<number> {
  const positions = await db.portfolioPosition.findMany();
  if (positions.length === 0) return 0;

  const sectorValues: Record<string, number> = {};
  let total = 0;

  for (const p of positions) {
    const price = p.currentPrice || Number(p.entryPrice);
    const mv = price * Number(p.quantity);
    const sector = p.sector || 'Unknown';
    sectorValues[sector] = (sectorValues[sector] || 0) + mv;
    total += mv;
  }

  if (total === 0) return 0;
  return round(
    Object.values(sectorValues).reduce((sum, v) => sum + ((v / total) * 100) ** 2, 0),
    2,
  );
}

export async function computeConcentration(prisma?: PrismaClient): Promise<Record<string, any>> {
  const db = prisma || getPrisma();
  const positions = await db.portfolioPosition.findMany();

  if (positions.length === 0) {
    return { top_3_pct: 0, top_5_pct: 0, largest_position: null, hhi: 0, position_weights: [] };
  }

  const holdings: { ticker: string; market_value: number; weight_pct: number }[] = [];
  let total = 0;

  for (const p of positions) {
    const price = p.currentPrice || Number(p.entryPrice);
    const mv = price * Number(p.quantity);
    holdings.push({ ticker: p.ticker, market_value: mv, weight_pct: 0 });
    total += mv;
  }

  if (total === 0) {
    return { top_3_pct: 0, top_5_pct: 0, largest_position: null, hhi: 0, position_weights: [] };
  }

  holdings.forEach(h => { h.weight_pct = round((h.market_value / total) * 100, 2); });
  holdings.sort((a, b) => b.market_value - a.market_value);

  const top3 = holdings.slice(0, 3).reduce((s, h) => s + h.weight_pct, 0);
  const top5 = holdings.slice(0, 5).reduce((s, h) => s + h.weight_pct, 0);
  const hhi = holdings.reduce((s, h) => s + h.weight_pct ** 2, 0);

  return {
    top_3_pct: round(top3, 2),
    top_5_pct: round(top5, 2),
    largest_position: holdings[0] ? { ticker: holdings[0].ticker, weight_pct: holdings[0].weight_pct } : null,
    hhi: round(hhi, 2),
    position_weights: holdings.map(h => ({ ticker: h.ticker, weight_pct: h.weight_pct })),
  };
}

export async function correlationClusters(
  prisma?: PrismaClient,
  threshold: number = 0.7,
): Promise<Record<string, any>> {
  const db = prisma || getPrisma();
  const corrDict = await buildCorrMatrix(db);

  if (Object.keys(corrDict).length === 0) {
    return { clusters: [], high_correlation_pairs: [] };
  }

  const tickers = Object.keys(corrDict);
  const n = tickers.length;

  if (n < 2) {
    return { clusters: [{ cluster_id: 1, tickers }], high_correlation_pairs: [] };
  }

  // Simple single-linkage clustering
  const clusters = tickers.map((t, i) => [i]);
  const findCluster = (idx: number) => clusters.findIndex(c => c.includes(idx));

  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      const corr = Math.abs(corrDict[tickers[i]]?.[tickers[j]] || 0);
      if (corr >= threshold) {
        const ci = findCluster(i);
        const cj = findCluster(j);
        if (ci !== cj && ci !== -1 && cj !== -1) {
          clusters[ci].push(...clusters[cj]);
          clusters.splice(cj, 1);
        }
      }
    }
  }

  const resultClusters = clusters
    .filter(c => c.length > 0)
    .map((c, idx) => ({
      cluster_id: idx + 1,
      tickers: c.map(i => tickers[i]),
    }));

  // High-correlation pairs
  const highPairs: any[] = [];
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      const corr = corrDict[tickers[i]]?.[tickers[j]] || 0;
      if (Math.abs(corr) >= threshold) {
        highPairs.push({
          ticker_1: tickers[i],
          ticker_2: tickers[j],
          correlation: round(corr, 4),
        });
      }
    }
  }

  return { clusters: resultClusters, high_correlation_pairs: highPairs };
}

export async function stressTestScenarios(
  prisma?: PrismaClient,
  drops?: number[],
): Promise<any[]> {
  const db = prisma || getPrisma();
  if (!drops) drops = [-10, -20, -30];

  const positions = await db.portfolioPosition.findMany();
  if (positions.length === 0) return [];

  const { returns } = await getPortfolioReturns(db, TRADING_DAYS_PER_YEAR);
  let portfolioBeta = await computeBeta(db, returns);
  if (portfolioBeta === null) portfolioBeta = 1.0;

  let totalValue = 0;
  const posData: { ticker: string; market_value: number }[] = [];

  for (const p of positions) {
    const price = p.currentPrice || Number(p.entryPrice);
    const mv = price * Number(p.quantity);
    totalValue += mv;
    posData.push({ ticker: p.ticker, market_value: mv });
  }

  return drops.map(dropPct => {
    const dropFrac = dropPct / 100;
    const portfolioImpactPct = portfolioBeta! * dropFrac;
    const portfolioImpactUsd = totalValue * portfolioImpactPct;

    const worstHit = posData
      .map(pd => ({ ticker: pd.ticker, impact_usd: round(pd.market_value * portfolioImpactPct, 2) }))
      .sort((a, b) => a.impact_usd - b.impact_usd)
      .slice(0, 3);

    return {
      label: `S&P 500 ${dropPct > 0 ? '+' : ''}${dropPct}%`,
      spy_drop_pct: dropPct,
      portfolio_impact_pct: round(portfolioImpactPct * 100, 2),
      portfolio_impact_usd: round(portfolioImpactUsd, 2),
      worst_hit_positions: worstHit,
    };
  });
}
